File: frontend/src/lib/paymentTerms.ts
const SPLIT_REGEX = /[\/,\-\s]+/;

/** Mensagem padrão quando a condição não é interpretável ou usa termos indefinidos (pedido de compra). */
export const MENSAGEM_CONDICAO_PAGAMENTO_PEDIDO_DEFINIDA =
  'Informe uma condição de pagamento definida. Ex.: À vista, 30, 30/45/60.';

function normalize(text: string): string {
  return (text ?? '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/ddl/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

/** Remove sufixos como "30 dias" -> "30" (uma parcela). */
function stripDiaSuffixes(normalized: string): string {
  return normalized.replace(/\b(\d+)\s*dias?\b/g, '$1').replace(/\s+/g, ' ').trim();
}

function isIndeterminatePayment(normalized: string): boolean {
  if (!normalized) return false;
  const markers = [
    'a combinar',
    'sob consulta',
    'apos aprovacao',
    'apos aprovacao do pedido',
    'apos pagamento',
    'apos pagamento do pedido',
    'a definir',
    'negociar',
    'conforme negocio',
    'conforme negociacao',
  ];
  return markers.some((m) => normalized.includes(m));
}

export type ParseCondicaoPagamentoResult =
  | { kind: 'dias'; dias: number[] }
  | { kind: 'invalid'; message: string };

/** Pedido de compra: exige condição fechada; termos indefinidos são inválidos. */
export function parseCondicaoPagamentoPedido(text: string): ParseCondicaoPagamentoResult {
  const raw = (text ?? '').trim();
  if (!raw) return { kind: 'dias', dias: [] };
  const normalized = normalize(raw);
  if (isIndeterminatePayment(normalized)) {
    return { kind: 'invalid', message: MENSAGEM_CONDICAO_PAGAMENTO_PEDIDO_DEFINIDA };
  }
  try {
    return { kind: 'dias', dias: parsePaymentCondition(text) };
  } catch {
    return { kind: 'invalid', message: MENSAGEM_CONDICAO_PAGAMENTO_PEDIDO_DEFINIDA };
  }
}

export function parsePaymentCondition(text: string): number[] {
  const raw = (text ?? '').trim();
  if (!raw) return [];

  const normalized = stripDiaSuffixes(normalize(raw));
  if (isIndeterminatePayment(normalized)) {
    return [];
  }
  if (normalized === 'a vista' || normalized === 'avista' || normalized === '0') {
    return [0];
  }

  const parts = normalized.split(SPLIT_REGEX).filter(Boolean);
  if (!parts.length) throw new Error("Condição inválida. Use formatos como '30/45 DDL' ou 'à vista'.");

  const days = parts.map((part) => {
    if (!/^\d+$/.test(part)) throw new Error('Condição inválida. Use apenas números e separadores.');
    const value = Number(part);
    if (value < 0) throw new Error('Dias da condição devem ser maiores ou iguais a zero.');
    return value;
  });

  return days;
}

export function buildDueDates(baseDate: string, days: number[]): string[] {
  if (!baseDate) return [];
  const base = new Date(`${baseDate}T00:00:00`);
  if (Number.isNaN(base.getTime())) return [];
  return days.map((day) => {
    const due = new Date(base);
    due.setDate(due.getDate() + day);
    return due.toISOString().slice(0, 10);
  });
}

/** Formata yyyy-mm-dd (ou ISO) para dd/mm/aaaa sem depender de fuso. */
export function formatDataIsoParaBr(iso: string): string {
  const s = (iso || '').slice(0, 10);
  if (!/^\d{4}-\d{2}-\d{2}$/.test(s)) return iso || '—';
  const [y, m, d] = s.split('-');
  return `${d}/${m}/${y}`;
}

File: backend/apps/comercial/payment_terms.py
import re
import unicodedata
from datetime import date, timedelta

from rest_framework import serializers


_SPLIT_PATTERN = re.compile(r"[/,\-\s]+")


def _normalize_text(value: str) -> str:
    normalized = unicodedata.normalize("NFKD", value or "")
    normalized = "".join(ch for ch in normalized if not unicodedata.combining(ch))
    return normalized.strip().lower()


def _is_indeterminate_payment(normalized: str) -> bool:
    """Condições em que não calculamos vencimentos nem quantidade de parcelas a partir de dias."""
    if not normalized:
        return False
    markers = (
        "a combinar",
        "a combinar.",
        "sob consulta",
        "apos aprovacao",
        "apos aprovacao do pedido",
        "apos pagamento",
        "apos pagamento do pedido",
        "a definir",
        "negociar",
        "conforme negocio",
        "conforme negociacao",
    )
    return any(m in normalized for m in markers)


def pagamento_integralmente_a_vista(dias_parcelas: list[int] | None) -> bool:
    """
    Plano canônico de pagamento integralmente à vista.

    Fonte: ArrayField `dias_parcelas` (inteiros), nunca texto livre.
    True somente quando há exatamente uma parcela com 0 dias: ``[0]``.

    Não classifica como à vista:
    - lista vazia / condição ausente (``[]``);
    - a prazo (ex.: ``[30]``, ``[30, 60]``);
    - misto entrada + prazo (ex.: ``[0, 30]``).

    Não define meio de pagamento fiscal (``tPag``) — apenas o prazo/plano.
    """
    days = list(dias_parcelas or [])
    return len(days) == 1 and int(days[0]) == 0


def parse_payment_condition(text: str) -> list[int]:
    raw = (text or "").strip()
    if raw == "":
        return []

    normalized = _normalize_text(raw)
    normalized = normalized.replace("ddl", " ")
    normalized = re.sub(r"\s+", " ", normalized).strip()

    if _is_indeterminate_payment(normalized):
        return []

    if normalized in {"a vista", "avista", "0"}:
        return [0]

    parts = [p for p in _SPLIT_PATTERN.split(normalized) if p]
    if not parts:
        raise serializers.ValidationError(
            "Condição inválida. Use formatos como '30 DDL', '30/45 DDL', '30/60/90' ou 'à vista'."
        )

    days: list[int] = []
    for part in parts:
        if not part.isdigit():
            raise serializers.ValidationError(
                "Condição inválida. Informe somente números e separadores (ex.: 30/45 DDL)."
            )
        n = int(part)
        if n < 0:
            raise serializers.ValidationError("Dias da condição de pagamento devem ser maiores ou iguais a zero.")
        days.append(n)

    if not days:
        raise serializers.ValidationError("Condição de pagamento vazia.")

    return days


def compute_due_dates(base_date: date, days: list[int]) -> list[date]:
    if base_date is None:
        return []
    return [base_date + timedelta(days=d) for d in days]
